[build] assertRepositoryURIUsesHttps breaks s3 repositories

When building a custom plugin using the build-tools jar, we get the following exception when building the 6.8.0 version:

```
Caused by: java.net.MalformedURLException: unknown protocol: s3
        at java_net_URI$toURL.call(Unknown Source)
        at org.elasticsearch.gradle.BuildPlugin.assertRepositoryURIUsesHttps(BuildPlugin.groovy:598)
        at jdk.internal.reflect.GeneratedMethodAccessor342.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.elasticsearch.gradle.BuildPlugin$_configureRepositories_closure11.doCall(BuildPlugin.groovy:567)
        at jdk.internal.reflect.GeneratedMethodAccessor341.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.util.ClosureBackedAction.execute(ClosureBackedAction.java:70)
        at org.gradle.util.ConfigureUtil.configureTarget(ConfigureUtil.java:154)
        at org.gradle.util.ConfigureUtil.configure(ConfigureUtil.java:105)
        at org.gradle.util.ConfigureUtil$WrappedConfigureAction.execute(ConfigureUtil.java:166)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1$1.run(DefaultCollectionCallbackActionDecorator.java:100)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:58)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:97)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        ... 173 more
```

This is because our build.gradle file has a repositories block like so:

```
    repositories {
        maven {
            url "s3://...
       }
    }
```

--- a/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
+++ b/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
@@ -48,7 +48,6 @@ import org.gradle.api.artifacts.ModuleVersionIdentifier
 import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.artifacts.ResolvedArtifact
 import org.gradle.api.artifacts.dsl.RepositoryHandler
-import org.gradle.api.artifacts.repositories.ArtifactRepository
 import org.gradle.api.artifacts.repositories.IvyArtifactRepository
 import org.gradle.api.artifacts.repositories.IvyPatternRepositoryLayout
 import org.gradle.api.artifacts.repositories.MavenArtifactRepository
@@ -85,12 +84,10 @@ import org.gradle.util.GradleVersion
 
 import java.nio.charset.StandardCharsets
 import java.nio.file.Files
-import java.time.ZoneOffset
-import java.time.ZonedDateTime
 import java.util.regex.Matcher
 
-import static org.elasticsearch.gradle.tool.Boilerplate.maybeConfigure
 import static org.elasticsearch.gradle.tool.Boilerplate.findByName
+import static org.elasticsearch.gradle.tool.Boilerplate.maybeConfigure
 
 /**
  * Encapsulates build configuration for elasticsearch projects.
@@ -411,11 +408,11 @@ class BuildPlugin implements Plugin<Project> {
         project.getRepositories().all { repository ->
             if (repository instanceof MavenArtifactRepository) {
                 final MavenArtifactRepository maven = (MavenArtifactRepository) repository
-                assertRepositoryURIUsesHttps(maven, project, maven.getUrl())
-                repository.getArtifactUrls().each { uri -> assertRepositoryURIUsesHttps(maven, project, uri) }
+                assertRepositoryURIIsSecure(maven.name, project.path, maven.getUrl())
+                repository.getArtifactUrls().each { uri -> assertRepositoryURIIsSecure(maven.name, project.path, uri) }
             } else if (repository instanceof IvyArtifactRepository) {
                 final IvyArtifactRepository ivy = (IvyArtifactRepository) repository
-                assertRepositoryURIUsesHttps(ivy, project, ivy.getUrl())
+                assertRepositoryURIIsSecure(ivy.name, project.path, ivy.getUrl())
             }
         }
         RepositoryHandler repos = project.repositories
@@ -455,9 +452,15 @@ class BuildPlugin implements Plugin<Project> {
         }
     }
 
-    private static void assertRepositoryURIUsesHttps(final ArtifactRepository repository, final Project project, final URI uri) {
-        if (uri != null && uri.toURL().getProtocol().equals("http")) {
-            throw new GradleException("repository [${repository.name}] on project with path [${project.path}] is using http for artifacts on [${uri.toURL()}]")
+    static void assertRepositoryURIIsSecure(final String repositoryName, final String projectPath, final URI uri) {
+        if (uri != null && ["file", "https", "s3"].contains(uri.getScheme()) == false) {
+            final String message = String.format(
+                    Locale.ROOT,
+                    "repository [%s] on project with path [%s] is not using a secure protocol for artifacts on [%s]",
+                    repositoryName,
+                    projectPath,
+                    uri.toURL())
+            throw new GradleException(message)
         }
     }
 

--- a/buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java
+++ b/buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java
@@ -45,6 +45,11 @@ public class BuildPluginTests extends GradleUnitTestCase {
         BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
     }
 
+    public void testRepositoryThatUsesFileScheme() throws URISyntaxException {
+        final URI uri = new URI("file:/tmp/maven");
+        BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
+    }
+
     public void testRepositoryURIThatUsesHttpsScheme() throws URISyntaxException {
         final URI uri = new URI("https://s3.amazonaws.com/artifacts.elastic.co/maven");
         BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);

--- a/buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java
+++ b/buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java
@@ -22,6 +22,9 @@ import org.elasticsearch.gradle.test.GradleUnitTestCase;
 import org.gradle.api.GradleException;
 import org.junit.Test;
 
+import java.net.URI;
+import java.net.URISyntaxException;
+
 
 public class BuildPluginTests extends GradleUnitTestCase {
 
@@ -34,6 +37,22 @@ public class BuildPluginTests extends GradleUnitTestCase {
     @Test(expected = GradleException.class)
     public void testFailingDockerVersions() {
         BuildPlugin.checkDockerVersionRecent("Docker version 17.04.0, build e68fc7a");
+    }
+
+    @Test(expected = GradleException.class)
+    public void testRepositoryURIThatUsesHttpScheme() throws URISyntaxException {
+        final URI uri = new URI("http://s3.amazonaws.com/artifacts.elastic.co/maven");
+        BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
+    }
+
+    public void testRepositoryURIThatUsesHttpsScheme() throws URISyntaxException {
+        final URI uri = new URI("https://s3.amazonaws.com/artifacts.elastic.co/maven");
+        BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
+    }
+
+    public void testRepositoryURIThatUsesS3Scheme() throws URISyntaxException {
+        final URI uri = new URI("s3://artifacts.elastic.co/maven");
+        BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
     }
 
 }

--- a/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
+++ b/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
@@ -18,18 +18,28 @@
  */
 package org.elasticsearch.gradle;
 
+import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.IOUtils;
 import org.elasticsearch.gradle.test.GradleIntegrationTestCase;
 import org.gradle.testkit.runner.BuildResult;
+import org.gradle.testkit.runner.GradleRunner;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
 
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
 
 public class BuildPluginIT extends GradleIntegrationTestCase {
+
+    @Rule
+    public TemporaryFolder tmpDir = new TemporaryFolder();
 
     public void testPluginCanBeApplied() {
         BuildResult result = getGradleRunner("elasticsearch.build")
@@ -44,6 +54,31 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
             .withArguments("check", "assemble", "-s", "-Dlocal.repo.path=" + getLocalTestRepoPath())
             .build();
         assertTaskSuccessful(result, ":check");
+    }
+
+    public void testInsecure() throws IOException {
+        final File projectDir = getProjectDir("elasticsearch.build");
+        FileUtils.copyDirectory(projectDir, tmpDir.getRoot(), pathname -> pathname.getPath().contains("/build/") == false);
+        final List<String> buildGradleLines =
+            Files.readAllLines(tmpDir.getRoot().toPath().resolve("build.gradle"), StandardCharsets.UTF_8);
+        // add an insecure repository to the build.gradle
+        final String url = "http://s3.amazonaws.com/artifacts.elastic.co/maven";
+        buildGradleLines.addAll(Arrays.asList(
+            "repositories {",
+                "  maven {",
+                "    name \"elastic-maven\"",
+                "    url \"" +  url + "\"\n",
+                "  }",
+                "}"));
+        Files.write(tmpDir.getRoot().toPath().resolve("build.gradle"), buildGradleLines, StandardCharsets.UTF_8);
+        final BuildResult result = GradleRunner.create()
+            .withProjectDir(tmpDir.getRoot())
+            .withArguments("clean", "hello", "-s", "-i", "--warning-mode=all", "--scan")
+            .withPluginClasspath()
+            .buildAndFail();
+        assertOutputContains(
+            result.getOutput(),
+            "repository [elastic-maven] on project with path [:] is not using a secure protocol for artifacts on [" + url + "]");
     }
 
     public void testLicenseAndNotice() throws IOException {

--- a/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
+++ b/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
@@ -56,20 +56,40 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
         assertTaskSuccessful(result, ":check");
     }
 
-    public void testInsecureArtifactRepository() throws IOException {
+    public void testInsecureMavenRepository() throws IOException {
+        final String name = "elastic-maven";
+        final String url = "http://s3.amazonaws.com/artifacts.elastic.co/maven";
+        // add an insecure maven repository to the build.gradle
+        final List<String> lines = Arrays.asList(
+            "repositories {",
+            "  maven {",
+            "    name \"elastic-maven\"",
+            "    url \"" +  url + "\"\n",
+            "  }",
+            "}");
+        runInsecureArtifactRepositoryTest(name, url, lines);
+    }
+
+    public void testInsecureIvyRepository() throws IOException {
+        final String name = "elastic-ivy";
+        final String url = "http://s3.amazonaws.com/artifacts.elastic.co/ivy";
+        // add an insecure ivy repository to the build.gradle
+        final List<String> lines = Arrays.asList(
+            "repositories {",
+            "  maven {",
+            "    name \"elastic-ivy\"",
+            "    url \"" +  url + "\"\n",
+            "  }",
+            "}");
+        runInsecureArtifactRepositoryTest(name, url, lines);
+    }
+
+    private void runInsecureArtifactRepositoryTest(final String name, final String url, final List<String> lines) throws IOException {
         final File projectDir = getProjectDir("elasticsearch.build");
         FileUtils.copyDirectory(projectDir, tmpDir.getRoot(), pathname -> pathname.getPath().contains("/build/") == false);
         final List<String> buildGradleLines =
             Files.readAllLines(tmpDir.getRoot().toPath().resolve("build.gradle"), StandardCharsets.UTF_8);
-        // add an insecure artifact repository to the build.gradle
-        final String url = "http://s3.amazonaws.com/artifacts.elastic.co/maven";
-        buildGradleLines.addAll(Arrays.asList(
-            "repositories {",
-                "  maven {",
-                "    name \"elastic-maven\"",
-                "    url \"" +  url + "\"\n",
-                "  }",
-                "}"));
+        buildGradleLines.addAll(lines);
         Files.write(tmpDir.getRoot().toPath().resolve("build.gradle"), buildGradleLines, StandardCharsets.UTF_8);
         final BuildResult result = GradleRunner.create()
             .withProjectDir(tmpDir.getRoot())
@@ -78,7 +98,7 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
             .buildAndFail();
         assertOutputContains(
             result.getOutput(),
-            "repository [elastic-maven] on project with path [:] is not using a secure protocol for artifacts on [" + url + "]");
+            "repository [" + name + "] on project with path [:] is not using a secure protocol for artifacts on [" + url + "]");
     }
 
     public void testLicenseAndNotice() throws IOException {

--- a/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
+++ b/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
@@ -56,7 +56,7 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
         assertTaskSuccessful(result, ":check");
     }
 
-    public void testInsecure() throws IOException {
+    public void testInsecureMavenRepository() throws IOException {
         final File projectDir = getProjectDir("elasticsearch.build");
         FileUtils.copyDirectory(projectDir, tmpDir.getRoot(), pathname -> pathname.getPath().contains("/build/") == false);
         final List<String> buildGradleLines =

--- a/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
+++ b/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
@@ -76,7 +76,7 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
         // add an insecure ivy repository to the build.gradle
         final List<String> lines = Arrays.asList(
             "repositories {",
-            "  maven {",
+            "  ivy {",
             "    name \"elastic-ivy\"",
             "    url \"" +  url + "\"\n",
             "  }",

--- a/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
+++ b/buildSrc/src/test/java/org/elasticsearch/gradle/BuildPluginIT.java
@@ -56,12 +56,12 @@ public class BuildPluginIT extends GradleIntegrationTestCase {
         assertTaskSuccessful(result, ":check");
     }
 
-    public void testInsecureMavenRepository() throws IOException {
+    public void testInsecureArtifactRepository() throws IOException {
         final File projectDir = getProjectDir("elasticsearch.build");
         FileUtils.copyDirectory(projectDir, tmpDir.getRoot(), pathname -> pathname.getPath().contains("/build/") == false);
         final List<String> buildGradleLines =
             Files.readAllLines(tmpDir.getRoot().toPath().resolve("build.gradle"), StandardCharsets.UTF_8);
-        // add an insecure repository to the build.gradle
+        // add an insecure artifact repository to the build.gradle
         final String url = "http://s3.amazonaws.com/artifacts.elastic.co/maven";
         buildGradleLines.addAll(Arrays.asList(
             "repositories {",

